### PR TITLE
Istio: improve network mode

### LINKED ISSUE #85: improve Istio network mode
address comment https://github.com/kurator-dev/kurator/pull/81#discussion_r980708368
after https://github.com/kurator-dev/kurator/pull/81 , we can customize cluster network when installing istio
we should remove `--network-mode`, let user define cluster network with labels on `Cluster` CR.

### PR DESCRIPTION
**Pull request by zirain**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #85

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: docs/content/en/docs/Components/istio.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+---
+title: "Integrate Istio with Kurator"
+linkTitle: "Integrate Istio with Kurator"
+description: >
+    Integrate Istio with Kurator?
+---
+
+### What is Istio
+
+[Istio](https://istio.io) is an open source service mesh that layers transparently onto existing distributed applications.
```
Comment (hzxuzhonghu) [inline docs/content/en/docs/Components/istio.md:10]: This is not a right sentence.

Istio is an open source service mesh that simplifies observability, traffic management, security, and policy for distributed applications.

File: docs/content/en/docs/Components/istio.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+In this task, we will show you how to install Istio with kurator.
+
+### Prerequisites
+
+{{% readfile "prerequisites-karmada.md" %}}
+
+### Install Istio
+
+Kurator provides a very simple command to install Istio [Primary-Remote](https://istio.io/latest/docs/setup/install/multicluster/primary-remote/) and add karmada-apiserver as the destination to apply [Istio configurations](https://istio.io/latest/docs/reference/config/) to.
```
Comment (hzxuzhonghu) [inline docs/content/en/docs/Components/istio.md:20]: Istio multicluster [Primary-Remote] model

File: docs/content/en/docs/Components/istio.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+- `--primary` specifies the cluster where the istio control plane install.
+- `--remote` specifies the cluster name joined to istio.
```
Comment (hzxuzhonghu) [inline docs/content/en/docs/Components/istio.md:23]: the cluster names that are managed by istio

File: docs/content/en/docs/Components/istio.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+```bash
+kurator install istio --primary member1 --remote member2
+```
+
+### Install Istio on different networks
+
+Kurator also providers a simple way to install [Istio on different network](https://istio.io/latest/docs/setup/install/multicluster/primary-remote_multi-network/).
```
Comment (hzxuzhonghu) [inline docs/content/en/docs/Components/istio.md:31]: different networks

File: docs/content/en/docs/Components/istio.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+First, you need label cluster to describe network topology. The following command will label clusters with different networks:
+
+```shell
+kubectl label cluster member1 topology.istio.io/network=member1 --overwrite --kubeconfig=/etc/karmada/karmada-apiserver.config
```
Comment (hzxuzhonghu) [inline docs/content/en/docs/Components/istio.md:36]: topology.istio.io/network=network1

File: docs/content/en/docs/Components/istio.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+kubectl label cluster member2 topology.istio.io/network=member2 --overwrite --kubeconfig=/etc/karmada/karmada-apiserver.config
```
Comment (hzxuzhonghu) [inline docs/content/en/docs/Components/istio.md:37]: ditto

File: pkg/plugin/istio/install.go [hunk 7]
```diff
@@ -47,16 +47,23 @@ import (
...
-	remotePilotAddressServiceName = "istiod-elb"
-	eastwestgatewayServiceName    = "istio-eastwestgateway"
-	istioSystemNamespace          = "istio-system"
-	istioOperatorNamespace        = "istio-operator"
-	karmadaClusterNamespace       = "karmada-cluster"
+	eastwestgatewayServiceName = "istio-eastwestgateway"
+	istioSystemNamespace       = "istio-system"
+	istioOperatorNamespace     = "istio-operator"
+	karmadaClusterNamespace    = "karmada-cluster"
+	defaultNetworkName         = "network-primary"
```
Comment (hzxuzhonghu) [inline pkg/plugin/istio/install.go:54]: why call "network-primary",  how about make "default" or empty

File: pkg/plugin/istio/install.go [hunk 8]
```diff
@@ -140,21 +143,16 @@ func (p *IstioPlugin) overrideNamespaceIfNeeded() error {
...
 	logrus.Infof("Begin to create istio cacerts")
 
-	nn := types.NamespacedName{
-		Namespace: istioOperatorNamespace,
-		Name:      ca.CASecret,
-	}
-
-	s, err := p.KubeClient().CoreV1().Secrets(nn.Namespace).Get(context.TODO(), nn.Name, metav1.GetOptions{})
+	s, err := p.KubeClient().CoreV1().Secrets(caSecret.Namespace).Get(context.TODO(), caSecret.Name, metav1.GetOptions{})
 	if err == nil {
```
Comment (hzxuzhonghu) [inline pkg/plugin/istio/install.go:147]: s !=nil

File: pkg/plugin/istio/install.go [hunk 9]
```diff
@@ -470,6 +425,9 @@ func (p *IstioPlugin) installRemotes(remotePilotAddress string) error {
 
 	for _, remote := range p.args.Remotes {
 		logrus.Infof("Begin to install istio in cluster %s", remote)
+		if err := waitSecertReady(p.Client, p.options, remote, caSecret); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/plugin/istio/install.go:428]: why do we need to wait caCerts created?

File: pkg/plugin/istio/install.go [hunk 10]
```diff
@@ -496,45 +454,47 @@ func (p *IstioPlugin) installRemotes(remotePilotAddress string) error {
...
+		logrus.Infof("primary cluster %s use default network", cluster)
+		return defaultNetworkName
+	}
 
-func networkName(cluster string) string {
-	logrus.Warnf("fallback to generate network name, cluster %s", cluster)
-	return fmt.Sprintf("%s-network", cluster)
+	logrus.Infof("cluster %s use primary's network", cluster)
+	// return primary cluster's network if not sepcifies
+	return p.clusterNetwork(p.args.Primary)
```
Comment (hzxuzhonghu) [inline pkg/plugin/istio/install.go:473]: I could not understand this.

Comment (hzxuzhonghu) [reply pkg/plugin/istio/install.go:473]: IMO, if we make default network empty, we donot need this, right?

File: docs/content/en/docs/Components/istio.md (hunk 1 above)
Comment (zirain) [reply docs/content/en/docs/Components/istio.md:10]: Istio is a service mesh—a modernized service networking layer that provides a transparent and language-independent way to flexibly and easily automate application network functions.

File: pkg/plugin/istio/install.go (hunk 9 above)
Comment (zirain) [reply pkg/plugin/istio/install.go:428]: if istiod running before `istio-ca-certs` synced by `karmada`, istiod will self-signed a certs, member2 will fail to join.

File: pkg/plugin/istio/install.go (hunk 10 above)
Comment (zirain) [reply pkg/plugin/istio/install.go:473]: yes

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**